File: examples/python/grpc/bidirectional_streaming.py
"""Bidirectional Streaming gRPC Example - Real-Time Chat.

This example demonstrates bidirectional streaming where clients and server
exchange messages concurrently in real-time.

Use case: Chat applications, multiplayer games, collaborative editing

Run:
    python examples/python/grpc/bidirectional_streaming.py
"""

from __future__ import annotations

import asyncio
import json
from typing import TYPE_CHECKING

from spikard.grpc import GrpcHandler, GrpcRequest, GrpcResponse

if TYPE_CHECKING:
    from collections.abc import AsyncGenerator, AsyncIterator


class ChatHandler(GrpcHandler):
    """Handler for real-time chat with bidirectional streaming.

    Implements bidirectional streaming RPC where clients and server
    exchange chat messages concurrently. Server can send messages
    to clients as they arrive from other users.
    """

    def __init__(self) -> None:
        """Initialize chat handler with message history."""
        self.message_history: list[dict] = []

    async def handle_request(self, request: GrpcRequest) -> GrpcResponse:
        """Unary RPC - Send single chat message."""
        json.loads(request.payload)

        resp_data = {"status": "delivered", "message_id": len(self.message_history)}
        return GrpcResponse(payload=json.dumps(resp_data).encode())

    async def handle_bidi_stream(self, request_stream: AsyncIterator[GrpcRequest]) -> AsyncGenerator[GrpcResponse]:
        """Bidirectional streaming RPC - Real-time chat.

        Client streams:
            {"user": "Alice", "text": "Hello!", "timestamp": 1234567890}
            {"user": "Alice", "text": "How are you?", "timestamp": 1234567891}
            ...

        Server streams:
            {"from": "Bob", "text": "Hi Alice!", "timestamp": 1234567891}
            {"from": "Charlie", "text": "Welcome!", "timestamp": 1234567892}
            ...

        Messages flow bidirectionally and concurrently.
        """
        async for request in request_stream:
            msg_data = json.loads(request.payload)

            user = msg_data.get("user", "Anonymous")
            text = msg_data.get("text", "")
            msg_data.get("timestamp", 0)

            # Store message
            self.message_history.append(msg_data)

            # Echo back with server processing
            response_data = {
                "from": "Server",
                "text": f"Message received from {user}: {text}",
                "timestamp": asyncio.get_event_loop().time(),
                "message_count": len(self.message_history),
            }

            yield GrpcResponse(payload=json.dumps(response_data).encode())

            # Simulate processing delay
            await asyncio.sleep(0.05)


class CollaborativeEditorHandler(GrpcHandler):
    """Handler for collaborative document editing with operational transforms."""

    def __init__(self) -> None:
        """Initialize editor with document state."""
        self.document_content = ""
        self.operation_count = 0

    async def handle_bidi_stream(self, request_stream: AsyncIterator[GrpcRequest]) -> AsyncGenerator[GrpcResponse]:
        """Bidirectional streaming RPC - Collaborative editing.

        Client streams operations:
            {"type": "insert", "position": 0, "text": "Hello"}
            {"type": "delete", "position": 3, "length": 2}
            {"type": "replace", "position": 0, "length": 5, "text": "Hi"}

        Server streams acknowledgments and broadcasts:
            {"operation_id": 1, "status": "applied", "document_version": 1}
            {"broadcast": true, "operation": {...}, "from_user": "Alice"}
        """
        async for request in request_stream:
            op_data = json.loads(request.payload)

            op_type = op_data.get("type", "unknown")
            position = op_data.get("position", 0)
            self.operation_count += 1

            # Apply operation to document
            if op_type == "insert":
                text = op_data.get("text", "")
                self.document_content = self.document_content[:position] + text + self.document_content[position:]

            elif op_type == "delete":
                length = op_data.get("length", 0)
                self.document_content = self.document_content[:position] + self.document_content[position + length :]

            elif op_type == "replace":
                length = op_data.get("length", 0)
                text = op_data.get("text", "")
                self.document_content = (
                    self.document_content[:position] + text + self.document_content[position + length :]
                )

            # Send acknowledgment
            ack_data = {
                "operation_id": self.operation_count,
                "status": "applied",
                "document_version": self.operation_count,
                "current_content": self.document_content,
            }

            yield GrpcResponse(payload=json.dumps(ack_data).encode())


class MultiplayerGameHandler(GrpcHandler):
    """Handler for multiplayer game with real-time state synchronization."""

    def __init__(self) -> None:
        """Initialize game state."""
        self.player_positions: dict[str, dict] = {}
        self.game_tick = 0

    async def handle_bidi_stream(self, request_stream: AsyncIterator[GrpcRequest]) -> AsyncGenerator[GrpcResponse]:
        """Bidirectional streaming RPC - Multiplayer game.

        Client streams player actions:
            {"player_id": "player1", "action": "move", "x": 10, "y": 20}
            {"player_id": "player1", "action": "shoot", "direction": "north"}

        Server streams game state updates:
            {"tick": 123, "players": {...}, "events": [...]}
        """
        async for request in request_stream:
            action_data = json.loads(request.payload)

            player_id = action_data.get("player_id", "unknown")
            action = action_data.get("action", "none")

            # Update player state
            if action == "move":
                x = action_data.get("x", 0)
                y = action_data.get("y", 0)
                self.player_positions[player_id] = {"x": x, "y": y}

            elif action == "shoot":
                action_data.get("direction", "north")

            self.game_tick += 1

            # Broadcast game state update
            state_update = {
                "tick": self.game_tick,
                "players": self.player_positions,
                "timestamp": asyncio.get_event_loop().time(),
            }

            yield GrpcResponse(payload=json.dumps(state_update).encode())

            # Game tick delay
            await asyncio.sleep(0.1)


async def simulate_chat_messages() -> AsyncIterator[GrpcRequest]:
    """Simulate a client sending chat messages."""
    messages = [
        "Hello everyone!",
        "How's it going?",
        "Anyone here?",
        "This is a test message",
        "Goodbye!",
    ]

    for i, text in enumerate(messages):
        msg_data = {
            "user": "Alice",
            "text": text,
            "timestamp": 1700000000 + i,
        }

        yield GrpcRequest(
            service_name="chat.v1.ChatService",
            method_name="Chat",
            payload=json.dumps(msg_data).encode(),
            metadata={},
        )

        await asyncio.sleep(0.2)


async def simulate_editing_operations() -> AsyncIterator[GrpcRequest]:
    """Simulate a client editing a document."""
    operations = [
        {"type": "insert", "position": 0, "text": "Hello"},
        {"type": "insert", "position": 5, "text": " World"},
        {"type": "delete", "position": 5, "length": 6},
        {"type": "insert", "position": 5, "text": " Spikard"},
    ]

    for op in operations:
        yield GrpcRequest(
            service_name="editor.v1.EditorService",
            method_name="EditDocument",
            payload=json.dumps(op).encode(),
            metadata={},
        )

        await asyncio.sleep(0.3)


async def example_bidirectional_streaming() -> None:
    """Demonstrate bidirectional streaming with mock requests."""
    # Example 1: Chat
    chat_handler = ChatHandler()

    message_stream = simulate_chat_messages()
    async for response in chat_handler.handle_bidi_stream(message_stream):
        json.loads(response.payload)

    # Example 2: Collaborative editing
    editor_handler = CollaborativeEditorHandler()

    edit_stream = simulate_editing_operations()
    async for response in editor_handler.handle_bidi_stream(edit_stream):
        json.loads(response.payload)


if __name__ == "__main__":
    # Run examples
    asyncio.run(example_bidirectional_streaming())

File: examples/node/src/grpc/bidirectional-streaming.ts
/**
 * Bidirectional Streaming gRPC Example: Real-time Chat Service
 *
 * This example demonstrates bidirectional streaming RPC where both client
 * and server send streams of messages concurrently.
 *
 * Current Implementation Note:
 * Due to napi-rs limitations, bidirectional streaming uses a collection
 * pattern where the handler consumes all client messages and returns
 * a batch of responses. This is similar to combining client streaming
 * (consume input) with server streaming (pre-collect output).
 *
 * For true bidirectional streaming in production, consider:
 * 1. WebSocket connections for real-time bidirectional communication
 * 2. Server-Sent Events (SSE) with HTTP/2 for unidirectional streaming
 * 3. Callback-based patterns (future enhancement)
 *
 * @see crates/spikard-node/src/grpc/handler.rs for implementation details
 */

import type { GrpcRequest, GrpcResponse } from "@spikard/node";

// Mock protobuf types (in production, use generated proto files)
interface ChatMessage {
	userId: string;
	username: string;
	message: string;
	timestamp: number;
	roomId: string;
}

interface ChatResponse {
	messageId: string;
	status: "delivered" | "failed";
	timestamp: number;
	error?: string;
}

interface ChatStreamResponse {
	responses: ChatResponse[];
}

/**
 * Chat Service Handler
 *
 * Demonstrates bidirectional streaming pattern where the handler:
 * 1. Consumes incoming messages from client stream
 * 2. Processes each message
 * 3. Returns batch of responses (due to napi-rs limitations)
 *
 * In a true bidirectional implementation, responses would be yielded
 * incrementally as messages are processed.
 */
class ChatServiceHandler {
	private messageHistory: Map<string, ChatMessage[]> = new Map();

	/**
	 * Bidirectional Streaming RPC: Real-time chat
	 *
	 * Current implementation uses the collection pattern:
	 * - Consumes all incoming messages from requestStream
	 * - Processes each message
	 * - Returns all responses in a single batch
	 *
	 * @param requestStream AsyncIterator yielding GrpcRequest objects
	 * @returns Promise<GrpcResponse> with batched ChatResponse array
	 */
	async handleBidiStream(requestStream: AsyncIterator<GrpcRequest>): Promise<GrpcResponse> {
		console.log("[Bidirectional Streaming] Chat called");

		const responses: ChatResponse[] = [];
		let messageCount = 0;

		try {
			// Consume all incoming messages
			for await (const request of requestStream) {
				messageCount++;
				console.log(`  Received message ${messageCount}`);

				// Deserialize incoming message
				const chatMsg = JSON.parse(request.payload.toString("utf-8")) as ChatMessage;
				console.log(`    ${chatMsg.username}: ${chatMsg.message}`);

				// Process message and generate response
				try {
					// Store message in room history
					if (!this.messageHistory.has(chatMsg.roomId)) {
						this.messageHistory.set(chatMsg.roomId, []);
					}
					this.messageHistory.get(chatMsg.roomId)!.push(chatMsg);

					// Simulate message processing
					await this.delay(50);

					// Create successful response
					const response: ChatResponse = {
						messageId: this.generateMessageId(chatMsg),
						status: "delivered",
						timestamp: Date.now(),
					};

					responses.push(response);
					console.log(`    ✓ Delivered: ${response.messageId}`);

					// Simulate broadcasting to other users (conceptual)
					await this.broadcastToRoom(chatMsg);
				} catch (error) {
					// Handle message processing error
					const errorResponse: ChatResponse = {
						messageId: this.generateMessageId(chatMsg),
						status: "failed",
						timestamp: Date.now(),
						error: error instanceof Error ? error.message : "Unknown error",
					};

					responses.push(errorResponse);
					console.log(`    ✗ Failed: ${errorResponse.error}`);
				}
			}
		} catch (error) {
			console.error("Error consuming stream:", error);
			throw new Error("INTERNAL: Failed to process chat stream");
		}

		console.log(`\nProcessed ${messageCount} messages, returning ${responses.length} responses`);

		// Return all responses in a batch
		const result: ChatStreamResponse = { responses };
		const responsePayload = Buffer.from(JSON.stringify(result), "utf-8");

		return {
			payload: responsePayload,
			metadata: {
				"content-type": "application/json",
				"x-service": "chat-service",
				"x-message-count": String(messageCount),
				"x-response-count": String(responses.length),
			},
		};
	}

	/**
	 * Generate unique message ID
	 */
	private generateMessageId(msg: ChatMessage): string {
		return `${msg.roomId}-${msg.userId}-${msg.timestamp}`;
	}

	/**
	 * Simulate broadcasting message to room (conceptual)
	 */
	private async broadcastToRoom(msg: ChatMessage): Promise<void> {
		// In production, this would:
		// 1. Look up all connected users in the room
		// 2. Send the message to each user's stream
		// 3. Handle disconnected users

		// Simulate broadcast delay
		await this.delay(20);

		const roomUsers = this.messageHistory.get(msg.roomId)?.length ?? 1;
		console.log(`    📡 Broadcast to ${roomUsers} users in room ${msg.roomId}`);
	}

	/**
	 * Get message history for a room
	 */
	getRoomHistory(roomId: string): ChatMessage[] {
		return this.messageHistory.get(roomId) ?? [];
	}

	/**
	 * Utility: Delay for ms milliseconds
	 */
	private delay(ms: number): Promise<void> {
		return new Promise((resolve) => setTimeout(resolve, ms));
	}

	/**
	 * Generate sample chat messages
	 */
	static generateSampleMessages(count: number, roomId: string): ChatMessage[] {
		const users = [
			{ id: "user1", name: "Alice" },
			{ id: "user2", name: "Bob" },
			{ id: "user3", name: "Charlie" },
			{ id: "user4", name: "Diana" },
		];

		const messageTemplates = [
			"Hello everyone!",
			"How are you doing?",
			"Check out this cool feature!",
			"Anyone up for a game?",
			"Thanks for the help!",
			"That makes sense",
			"I agree with that",
			"Interesting point",
			"Could you clarify?",
			"See you later!",
		];

		const messages: ChatMessage[] = [];
		const baseTimestamp = Date.now() - 60000; // Start 1 minute ago

		for (let i = 0; i < count; i++) {
			const user = users[i % users.length];
			const message = messageTemplates[Math.floor(Math.random() * messageTemplates.length)];

			messages.push({
				userId: user.id,
				username: user.name,
				message,
				timestamp: baseTimestamp + i * 2000, // 2 seconds apart
				roomId,
			});
		}

		return messages;
	}
}

/**
 * Mock async generator to simulate client message stream
 */
async function* createChatStream(messages: ChatMessage[]): AsyncGenerator<GrpcRequest> {
	for (const msg of messages) {
		// Simulate typing and network delay
		await new Promise((resolve) => setTimeout(resolve, 100));

		yield {
			serviceName: "chat.v1.ChatService",
			methodName: "StreamChat",
			payload: Buffer.from(JSON.stringify(msg), "utf-8"),
			metadata: {
				"user-id": msg.userId,
				"room-id": msg.roomId,
			},
		};
	}
}

/**
 * Client: Test the chat service
 */
async function runClient() {
	console.log("=== Bidirectional Chat Client ===\n");

	const service = new ChatServiceHandler();
	const roomId = "general";

	// Generate sample chat messages
	const messages = ChatServiceHandler.generateSampleMessages(12, roomId);

	console.log(`Generated ${messages.length} chat messages for room: ${roomId}\n`);
	console.log("Streaming messages to chat service...\n");

	// Create client stream
	const chatStream = createChatStream(messages);

	// Call bidirectional streaming RPC
	const response = await service.handleBidiStream(chatStream);

	// Parse and display results
	const result = JSON.parse(response.payload.toString("utf-8")) as ChatStreamResponse;

	console.log("\n=== Chat Session Results ===");
	console.log(`Messages sent: ${messages.length}`);
	console.log(`Responses received: ${result.responses.length}`);

	// Display delivery status
	const delivered = result.responses.filter((r) => r.status === "delivered").length;
	const failed = result.responses.filter((r) => r.status === "failed").length;

	console.log(`\nDelivery status:`);
	console.log(`  ✓ Delivered: ${delivered}`);
	console.log(`  ✗ Failed: ${failed}`);

	// Display sample responses
	console.log(`\nSample responses:`);
	result.responses.slice(0, 5).forEach((resp, i) => {
		const icon = resp.status === "delivered" ? "✓" : "✗";
		const time = new Date(resp.timestamp).toISOString();
		console.log(`  ${icon} ${resp.messageId} @ ${time}`);
		if (resp.error) {
			console.log(`    Error: ${resp.error}`);
		}
	});

	if (result.responses.length > 5) {
		console.log(`  ... and ${result.responses.length - 5} more`);
	}

	// Display room history
	const history = service.getRoomHistory(roomId);
	console.log(`\nRoom history: ${history.length} messages stored`);

	console.log("\n=== Demo Complete ===");
}

/**
 * Server: Start the gRPC server (when Spikard supports it)
 */
async function runServer() {
	console.log("=== Starting Chat Service ===\n");

	// Note: This is a conceptual example. Actual server implementation
	// depends on Spikard's gRPC server API being finalized.
	//
	// Expected usage:
	// const app = new Spikard();
	// app.registerGrpcHandler('chat.v1.ChatService', new ChatServiceHandler());
	// await app.listen(50051);

	console.log("Server would run on port 50051");
	console.log("Service: chat.v1.ChatService");
	console.log("Methods: StreamChat (bidirectional streaming)\n");

	console.log("⚠️  Server implementation pending Spikard gRPC server API");
}

/**
 * Main entry point
 */
async function main() {
	const mode = process.argv[2] || "client";

	if (mode === "server") {
		await runServer();
	} else {
		await runClient();
	}
}

// Run if executed directly
if (require.main === module) {
	main().catch((error) => {
		console.error("Error:", error);
		process.exit(1);
	});
}

export { ChatServiceHandler, type ChatMessage, type ChatResponse, type ChatStreamResponse, createChatStream };
